src/gatherers/critical-network-chains.js

```diff
@@ -0,0 +1,149 @@
+
+  constructor(request, parent) {
+    // The children of a RequestNode are the requests initiated by it
+    this.children = [];
+    // The parent of a RequestNode is the request that initiated it
+    this.parent = parent;
+
+    this.request = request;
+  }
+
+  setParent(parentNode) {
+    this.parent = parentNode;
+  }
+
+  addChild(childNode) {
+    this.children.push(childNode);
+  }
+
+  toJSON() {
+    // Prevents circular reference so we can print nodes when needed
+    return `{
+      id: ${this.requestId},
+      parent: ${this.parent ? this.parent.requestId : null},
+      children: ${JSON.stringify(this.children.map(child => child.requestId))}
+    }`;
+  }
+
+}
+
+class CriticalNetworkChains extends Gather {
+
+  get criticalPriorities() {
+    // For now, critical request == render blocking request (as decided by
+    // blink). Blink treats requests with the following priority levels as
+    // render blocking.
+    // See https://docs.google.com/document/d/1bCDuq9H1ih9iNjgzyAL0gpwNFiEP4TZS-YLRp_RuMlc
+    return ['VeryHigh', 'High', 'Medium'];
+  }
+
+  getCriticalChains(networkRecords) {
+    // Drop the first request because it's uninteresting - it's the page html
+    // and always critical. No point including it in every request
+    const criticalRequests = networkRecords.slice(1).filter(
+      req => includes(this.criticalPriorities, req.initialPriority()));
+
+    // Build a map of requestID -> Node.
+    const requestIdToNodes = new Map();
+    for (let request of criticalRequests) {
+      const requestNode = new RequestNode(request, null);
+      requestIdToNodes.set(requestNode.requestId, requestNode);
+    }
+
+    // Connect the parents and children
+    for (let request of criticalRequests) {
+      if (request.initiatorRequest()) {
+        const parentRequestId = request.initiatorRequest().requestId;
+        const childNode = requestIdToNodes.get(request.requestId);
+        const parentNode = requestIdToNodes.get(parentRequestId);
+        if (childNode && parentNode) {
+          // Both child and parent must be critical
+          // TODO: We may need handle redirects carefully. Investigate
+          childNode.setParent(parentNode);
+          parentNode.addChild(childNode);
+        }
+      }
+    }
+
+    const nodesList = [...requestIdToNodes.values()];
+    const orphanNodes = nodesList.filter(node => node.parent === null);
+    const nodeChains = flatten(orphanNodes.map(
+      node => this._getChainsDFS(node)));
+    const requestChains = nodeChains.map(chain => chain.map(
+      node => node.request));
+    return requestChains;
+  }
+
+  postProfiling(options, tracingData) {
+    const chains = this.getCriticalChains(tracingData.networkRecords);
+
+    if (options.flags.useNetDepGraph) {
+      // There logs are here so we can test this gatherer
+      // Will be removed when we have a way to surface them in the report
+      const nonTrivialChains = chains.filter(chain => chain.length > 1);
```

Comment:
ah actually we want `.sort((a,b) => b.totalTimeBetweenBeginAndEnd - a.totalTimeBetweenBeginAndEnd)` but after the bit below. :)